### PR TITLE
Add LayoutLMv3 backbone

### PR DESCRIPTION
**Pull request by Bourn3z**
Thank you for your contribution to the MindOCR repo.
Before submitting this PR, please make sure:

- [ ] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved terminology
- [ ] You have added unit tests

## Motivation

增加LayoutLMv3主干网络，并适配下游SER以及RE任务。
Add LayoutLMv3 backbone and adapt to downstream SER and RE tasks.

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: README_CN.md
Comment (panshaowu) [inline README_CN.md:1]: 1. 需要在`What is New`清单中刷新新增特性日志；
2. 英文版本readme也需要刷新。

File: configs/kie/layoutlmv3/README_CN.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
+[English](README.md) | 中文
```
Comment (panshaowu) [inline configs/kie/layoutlmv3/README_CN.md:1]: 需要补充本页面下的英文版README.md，当前点击会404.

File: configs/kie/layoutlmv3/README_CN.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+# LayoutLMv3
+<!--- Guideline: use url linked to abstract in ArXiv instead of PDF for fast loading.  -->
+
+> [LayoutLMv3: Pre-training for Document AI with Unified Text and Image Masking](https://arxiv.org/abs/2204.08387)
+
+## 1. 模型描述
+<!--- Guideline: Introduce the model and architectures. Cite if you use/adopt paper explanation from others. -->
+
+不同于以往的LayoutLM系列模型，在模型架构设计上，LayoutLMv3 不依赖复杂的 CNN 或 Faster R-CNN 网络来表征图像，而是直接利用文档图像的图像块，从而大大节省了参数并避免了复杂的文档预处理（如人工标注目标区域框和文档目标检测）。简单的统一架构和训练目标使 LayoutLMv3 成为通用的预训练模型，可适用于以文本为中心和以图像为中心的文档 AI 任务。
```
Comment (panshaowu) [inline configs/kie/layoutlmv3/README_CN.md:11]: “LayoutLMv3 不依赖复杂的 CNN 或 Faster R-CNN 网络来表征图像，而是直接利用文档图像的图像块”的意思是否是“LayoutLMv3 不依赖前置的 CNN 或 Faster R-CNN 网络提取图像特征，而是直接将文档图像的图像块作为网络输入”

File: configs/kie/layoutlmv3/README_CN.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+在包括以文本为中心的数据集：表单理解 FUNSD 数据集，票据理解 CORD 数据集，文档视觉问答 DocVQA 数据集；以及以图像为中心的数据集：文档图像分类 RVL-CDIP 数据集，文档布局分析 PubLayNet 数据集。实验结果表明，LayoutLMv3 在这些数据集上以更少的参数达到了更优的性能。
```
Comment (panshaowu) [inline configs/kie/layoutlmv3/README_CN.md:13]: 请看下以下方式排版会否更清晰？

实验结果表明，LayoutLMv3 在以下数据集上以更少的参数达到了更优的性能：
1. 以文本为中心的数据集：表单理解 FUNSD 数据集，票据理解 CORD 数据集，文档视觉问答 DocVQA 数据集；
2. 以图像为中心的数据集：文档图像分类 RVL-CDIP 数据集，文档布局分析 PubLayNet 数据集。

File: configs/kie/layoutlmv3/README_CN.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+- Context: Training context denoted as {device}x{pieces}-{MS mode}, where mindspore mode can be G - graph mode or F - pynative mode with ms function. For example, D910x8-G is for training on 8 pieces of Ascend 910 NPU using graph mode.
+- Top-1 and Top-5: Keep 2 digits after the decimal point.
+- Params (M): # of model parameters in millions (10^6). Keep 2 digits after the decimal point
+- Recipe: Training recipe/configuration linked to a yaml config file. Use absolute url path.
+- Download: url of the pretrained model weights. Use absolute url path.
+-->
+
+### 训练端
+
+根据我们的实验，训练（[模型训练](#32-模型训练)）性能和精度评估（[模型评估](#33-模型评估)）结果如下：
```
Comment (panshaowu) [inline configs/kie/layoutlmv3/README_CN.md:41]: 此处的超链接引用是否有问题？

File: configs/kie/layoutlmv3/README_CN.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+#### 3.1.1 安装
+环境安装教程请参考MindOCR的 [installation instruction](https://github.com/mindspore-lab/mindocr#installation).
+
+#### 3.1.2 数据集下载
+这里使用[XFUND数据集](https://github.com/doc-analysis/XFUND)做为实验数据集。 XFUN数据集是微软提出的一个用于KIE任务的多语言数据集，共包含七个数据集，每个数据集包含149张训练集和50张验证集
+
+分别为：ZH(中文)、JA(日语)、ES(西班牙)、FR(法语)、IT(意大利)、DE(德语)、PT(葡萄牙)
+
+这里提供了经过预处理可以直接使用的[中文数据集](https://download.mindspore.cn/toolkits/mindocr/vi-layoutxlm/XFUND.tar)供大家下载。
```
Comment (panshaowu) [inline configs/kie/layoutlmv3/README_CN.md:93]: 经过预处理可以直接使用的
->
已经过预处理、可直接用于训练的

File: configs/kie/layoutlmv3/README_CN.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+### 3.3 模型评估
+
+若要评估已训练模型的准确性，可以使用`eval.py`。请在yaml配置文件的`eval`部分将参数`ckpt_load_path`设置为模型checkpoint的文件路径，然后运行：
+
+```
+python tools/eval.py --config configs/kie/layoutlmv3/ser_layoutlmv3_xfund_zh.yaml
+```
+
+
+## 4. MindSpore Lite 推理
```
Comment (panshaowu) [inline configs/kie/layoutlmv3/README_CN.md:245]: 由于后续mindspore版本将主推训推一体，此节可删除。

File: configs/kie/layoutlmv3/README_CN.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+**注意:**
+- 由于全局批大小 （batch_size x num_devices） 是对结果复现很重要，因此当GPU/NPU卡数发生变化时，调整`batch_size`以保持全局批大小不变，或根据新的全局批大小线性调整学习率。
+
+
+### 3.2 模型训练
+<!--- Guideline: Avoid using shell script in the command line. Python script preferred. -->
+* 分布式训练
+
+使用预定义的训练配置可以轻松重现报告的结果。对于在多个昇腾910设备上的分布式训练，请将配置参数`distribute`修改为True，并运行：
```
Comment (panshaowu) [inline configs/kie/layoutlmv3/README_CN.md:217]: 此处建议说明是单机多卡的数据并行训练，以避免误会是模型/张量并行训练。

File: configs/kie/layoutlmv3/README_CN.md (hunk 2 above)
Comment (Bourn3z) [reply configs/kie/layoutlmv3/README_CN.md:11]: 是的，我修改一下表述

File: configs/kie/layoutlmv3/README_CN.md (hunk 1 above)
Comment (Bourn3z) [reply configs/kie/layoutlmv3/README_CN.md:1]: 正在补充

File: README_CN.md
Comment (Bourn3z) [reply README_CN.md:1]: 已增加

File: mindocr/models/backbones/__init__.py [hunk 8]
```diff
@@ -16,7 +17,6 @@
 from .rec_svtr import *
 from .rec_svtr_enhance import *
 from .rec_vgg import *
-from .table_master_resnet import *
```
Comment (chengtianwu) [inline mindocr/models/backbones/__init__.py:19]: 这个删了，会影响到table master吗

File: mindocr/models/backbones/layoutlmv3/layoutlmv3.py [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+        output_hidden_states=None,
+        return_dict=None,
+    ):
+        r"""
+        Returns:
+
+        Examples:
+
+        ```python
+        >>> from transformers import AutoProcessor, AutoModel
```
Comment (chengtianwu) [inline mindocr/models/backbones/layoutlmv3/layoutlmv3.py:407]: 这个example干掉吧

File: mindocr/models/backbones/layoutlmv3/layoutlmv3.py [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+
+        if config.visual_embed:
+            # use the default pre-training parameters for fine-tuning (e.g., input_size)
+            # when the input_size is larger in fine-tuning, we will interpolate the position embeddings in forward
+            self.patch_embed = LayoutLMv3PatchEmbeddings(config)
+
+            size = int(config.input_size / config.patch_size)
+            self.cls_token = Parameter(ops.zeros((1, 1, config.hidden_size)))
+            self.pos_embed = Parameter(ops.zeros((1, size * size + 1, config.hidden_size)))
+            self.pos_drop = nn.Dropout(p=0.0)
```
Comment (chengtianwu) [inline mindocr/models/backbones/layoutlmv3/layoutlmv3.py:249]: 这个dropout是不是没用，可以删掉

File: mindocr/models/backbones/layoutlmv3/layoutlmv3.py [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+
+    def cogview_attention(self, attention_scores: Tensor, alpha=32):
+        """
+        https://arxiv.org/abs/2105.13290 Section 2.4 Stabilization of training: Precision Bottleneck Relaxation
+        (PB-Relax). A replacement of the original nn.Softmax(dim=-1)(attention_scores). Seems the new attention_probs
+        will result in a slower speed and a little bias. Can use torch.allclose(standard_attention_probs,
+        cogview_attention_probs, atol=1e-08) for comparison. The smaller atol (e.g., 1e-08), the better.
+        """
+        scaled_attention_scores = attention_scores / alpha
+        max_value = scaled_attention_scores.max(axis=(-1)).unsqueeze(-1)
```
Comment (chengtianwu) [inline mindocr/models/backbones/layoutlmv3/layoutlmv3.py:144]: 改成：
max_value = scaled_attention_scores.max(axis=-1).unsqueeze(-1)

File: configs/kie/layoutlmv3/ser_layoutlmv3_xfund_zh.yaml [hunk 12]
```diff
@@ -0,0 +1,128 @@
+system:
+  mode: 0 # 0 for graph mode, 1 for pynative mode in MindSpore
+  distribute: False
+  amp_level: "O0"
```
Comment (chengtianwu) [inline configs/kie/layoutlmv3/ser_layoutlmv3_xfund_zh.yaml:4]: 开O0会不会太慢了

File: mindocr/models/backbones/layoutlmv3/layoutlmv3.py (hunk 10 above)
Comment (Bourn3z) [reply mindocr/models/backbones/layoutlmv3/layoutlmv3.py:249]: 功能上是没用，但还是保留吧，毕竟是网络结构的一部分。

File: configs/kie/layoutlmv3/ser_layoutlmv3_xfund_zh.yaml (hunk 12 above)
Comment (Bourn3z) [reply configs/kie/layoutlmv3/ser_layoutlmv3_xfund_zh.yaml:4]: 全连接层都手动to_float16了，之前amp有bug。

**Review by chengtianwu — APPROVED** [review]

**Review by panshaowu — APPROVED** [review]

**Review by chengtianwu — APPROVED** [review]
